torch/csrc/jit/ir/alias_analysis.cpp

```diff
@@ -126,10 +126,81 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
 
 AliasDb::~AliasDb() = default;
 
+// Structure that keeps tracks of all writes at a high level. When alias
+// analysis is done, this will be used to construct a more efficient WriteIndex.
+struct AliasDb::WriteRegistry {
+  void registerWrite(const Value* v, Node* n) {
+    writes_[n].emplace_back(v);
+  }
+  void registerWriteToAllContained(const Value* v, Node* n) {
+    containedWrites_[n].emplace_back(v);
+  }
+  void registerWriteToAllWildcards(Node* n) {
+    writesToAllWildcards_.insert(n);
+  }
+  std::unordered_map<Node*, std::vector<const Value*>> writes_;
+  std::unordered_map<Node*, std::vector<const Value*>> containedWrites_;
+  std::unordered_set<Node*> writesToAllWildcards_;
+};
+
 AliasDb::AliasDb(std::shared_ptr<Graph> graph, bool isFrozen)
-    : graph_(std::move(graph)), isFrozen_(isFrozen) {
-  memoryDAG_ = torch::make_unique<MemoryDAG>();
+    : graph_(std::move(graph)),
+      isFrozen_(isFrozen),
+      memoryDAGBuilder_(std::make_unique<MemoryDAGBuilder>()),
+      writeRegistry_(std::make_unique<AliasDb::WriteRegistry>()) {
   analyze(graph_);
+
+  memoryDAG_ = std::make_unique<MemoryDAG>(std::move(memoryDAGBuilder_));
+  memoryDAGBuilder_ = nullptr; // to make further access a hard error
+
+  memoryDAG_->setWildcards(
+      wildcards_, elementMap_, [&](const Value* v) -> Element* {
+        return getWildcard(v->type());
+      });
+
+  // Now we build up the various write indices based on information in the write
+  // registry that we populated during analysis
+
+  // Initialize the write index
+  writeIndex_ = TWriteIndex();
+  auto& writeIndex = *writeIndex_; // to make operator[] less ugly
+
+  // build the write index
+  for (const auto& write : writeRegistry_->writes_) {
+    Node* node = write.first;
+    const std::vector<const Value*> writtenValues = write.second;
+    for (const Value* writtenValue : writtenValues) {
+      auto it = elementMap_.find(writtenValue);
+      TORCH_INTERNAL_ASSERT(
+          it != elementMap_.end(), "Tried to write to value not in MemoryDAG");
+      const auto& writtenMemoryLocations =
+          memoryDAG_->getMemoryLocations(it->second);
+      writeIndex[node] |= writtenMemoryLocations;
+    }
+  }
+
+  for (const auto& write : writeRegistry_->containedWrites_) {
+    Node* node = write.first;
+    const std::vector<const Value*> writtenValues = write.second;
+    for (const Value* writtenValue : writtenValues) {
+      auto elem = elementMap_.at(writtenValue);
+      MemoryLocations writtenMemoryLocations;
+      memoryDAG_->collectAllContainedMemoryLocations(
+          elem, writtenMemoryLocations);
+      writeIndex[node] |= writtenMemoryLocations;
+    }
+  }
+
+  for (const auto& write : writeRegistry_->writesToAllWildcards_) {
+    for (const auto& pr : wildcardIndex_) {
+      writeIndex[write].set(pr.second->index);
```

Comment:
are you sure this api call is right `writeIndex[write].set(pr.second->index)` ? Why are we using `.set()` here instead of `|=` as above ?

Reply:
|= is for adding many memorylocations. set() is for adding one.